token-swap/program/src/curve.rs

```diff
@@ -1,11 +1,93 @@
 //! Swap calculations and curve implementations
 
+use solana_sdk::program_error::ProgramError;
+use std::convert::TryFrom;
+
 /// Initial amount of pool tokens for swap contract, hard-coded to something
 /// "sensible" given a maximum of u64.
 /// Note that on Ethereum, Uniswap uses the geometric mean of all provided
 /// input amounts, and Balancer uses 100 * 10 ^ 18.
 pub const INITIAL_SWAP_POOL_AMOUNT: u64 = 1_000_000_000;
 
+/// Curve types supported by the token-swap program.
+#[repr(C)]
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum CurveType {
+    /// Uniswap-style constant product curve, invariant = token_a_amount * token_b_amount
+    ConstantProduct,
+    /// Flat line, always providing 1:1 from one token to another
+    Flat,
+}
+
+impl Default for CurveType {
+    fn default() -> Self {
+        CurveType::ConstantProduct
+    }
+}
+
+impl TryFrom<u8> for CurveType {
+    type Error = ProgramError;
+
+    fn try_from(curve_type: u8) -> Result<Self, Self::Error> {
+        match curve_type {
+            0 => Ok(CurveType::ConstantProduct),
+            1 => Ok(CurveType::Flat),
+            _ => Err(ProgramError::InvalidAccountData),
+        }
+    }
+}
+
+impl CurveType {
+    /// Create a swap curve corresponding to the enum
+    pub fn swap_curve(
+        &self,
+        swap_source_amount: u64,
+        swap_destination_amount: u64,
+        fee_numerator: u64,
+        fee_denominator: u64,
```

Comment:
Can you explain more how this will work for different curves with different inputs? My initial take is that this doesn't seem like the right abstraction, if the inputs will vary.

Reply:
You're totally right, I wasn't sure how to cover many different types of curves, especially if the curve itself requires instantiation parameters.  Just the `swap_source_amount` and `swap_destination_amount` should be the main inputs, and everything else is encoded in the curve, including fees and other parameters that might be needed.

Here's what I propose: 
* `SwapInfo` includes some extra space (64 bytes?) for the curve parameters, also passed on init
* `SwapCurve` implements `Pack`
* include `deposit` and `withdraw` in the `SwapCurve` interface -- for example, some extra parameters are needed for stable swap.
* the call to `swap` passes in `swap_source_amount` and `swap_destination_amount`

what do you think?

Reply:
The `Pack` implementation is to easily go from `SwapCurve` to the parameters stored in `SwapInfo` and back? If I've got that right, this sounds pretty good to me!

Reply:
Exactly, I'll do that then!